std/experimental/logger/asynclogger.d

```diff
@@ -0,0 +1,161 @@
+-------------
+void main()
+{
+    import std.algorithm;
+    import std.conv;
+    import std.stdio;
+    import std.datetime;
+    import std.experimental.logger;
+
+    auto fl = new FileLogger("defaultFileLogger.log", LogLevel.all);
+    void logFileDefault()
+    {
+        fl.log("test message");
+    }
+
+    auto afl = new AsyncLogger(new FileLogger("asyncFileLogger.log", LogLevel.all));
+    void logFileAsync()
+    {
+        afl.log("test message");
+    }
+
+    auto r = benchmark!(logFileDefault, logFileAsync)(100_000);
+    writeln(map!(a => to!string(to!Duration(a)))(r[]));
+}
+-------------
+*/
+class AsyncLogger : Logger
+{
+    /// Shuntdown message
+    struct ShutdownMsg {}
+
+    import std.datetime : SysTime;
+
+    /** LogEntryMsg is an aggregation copying original $(D LogEntry)
+    but $(D (hasUnsharedAliasing!LogEntryMsg == true)).
+    This aggregation will be send to logging thread.
+    */
+    protected struct LogEntryMsg
+    {
+        /// the filename the log function was called from
+        string file;
+        /// the line number the log function was called from
+        int line;
+        /// the name of the function the log function was called from
+        string funcName;
+        /// the pretty formatted name of the function the log function was
+        /// called from
+        string prettyFuncName;
+        /// the name of the module the log message is coming from
+        string moduleName;
+        /// the $(D LogLevel) associated with the log message
+        LogLevel logLevel;
+        /// the time the message was logged
+        SysTime timestamp;
+        /// the message of the log message
+        string msg;
+    }
+
+    private static void spawnedFunc(shared Logger ptr) @trusted
+    {
+        auto l = cast(Logger)(ptr);
+        bool isRunning = true;
+        while (isRunning)
+        {
+            receive(
+                (LogEntryMsg e)
+                {
+                    LogEntry tmp = LogEntry(e.file, e.line, e.funcName, e.prettyFuncName, e.moduleName, e.logLevel, ownerTid, e.timestamp, e.msg);
+                    l.forwardMsg(tmp);
+                },
+                (ShutdownMsg msg) { isRunning = false; },
+                (Exception ex) { isRunning = false; },
+                (Variant v) {}
+            );
+        }
+    }
+
+    /** This constructor sets a LogLevel and starts a new logging thread
+    that logs every entry asynchronously by using Logger. Because logger writes down
+    all entries into special logging thread, logger should not be used further
+    (without confidence that logger is not being used at the moment).
+    */
+    this(Logger logger, LogLevel lv = LogLevel.all) @trusted
+    {
+        super(lv);
+        _logger = cast(shared Logger)(logger);
+        _tid = spawn(&(spawnedFunc), _logger);
+    }
+
+    /** Returns the Tid of logging thread.
+    It can be useful for setting mailbox size, crowding rules, etc.
+    */
+    Tid tid() @safe
```

Comment:
Yes, let's make the class final to begin with and hide away tid if possible.
We can lift restrictions later on, iff necessary, adding them is much harder.